### PR TITLE
Add dripsHistoryHashes to SqueezedDrips event

### PR DESCRIPTION
**Pull request by CodeSandwich**
Fixes https://github.com/radicle-dev/drips-contracts/issues/242.

### CODE REVIEW

File: src/Drips.sol [hunk 1]
```diff
@@ -337,17 +345,24 @@ abstract contract Drips {
         DripsHistory[] memory dripsHistory
     ) internal returns (uint128 amt) {
         uint256 squeezedNum;
-        uint256[] memory squeezedIdxs;
-        (amt, squeezedNum, squeezedIdxs) =
+        uint256[] memory squeezedRevIdxs;
+        bytes32[] memory historyHashes;
+        uint256 currCycleConfigs;
+        (amt, squeezedNum, squeezedRevIdxs, historyHashes, currCycleConfigs) =
```
Comment (xmxanuel) [inline src/Drips.sol:351]: It is kind of interesting with how many different arrays we are dealing in these two functions. The complexity is really high.

1. `dripsHistory` array
   - includes all the information
2.  `squeezedHistoryHashes`
   - only the hashes from the provided dripsHistory
3. `nextSqueezed` array with` currCycleConfigs` as length
   - array can contain elements from outdated cycles :)
4. `squeezedRevIdxs` array with `squeezedNum` as length
   - basically how often we touched `nextSqueezed`
   - hashes in reverse order
5. `squeezedHistoryHashes` array for events length `squeezedNum`
   - array we construct only the present the hashes in the right order


Then it is important to note that we can have three different length
`squeezedNum != historyHashes.length != currCycleConfigs`.

(They can also be the same :) )

File: src/Drips.sol [hunk 2]
```diff
@@ -359,34 +374,55 @@ abstract contract Drips {
...
-        squeezedIdxs = new uint256[](historyIdx);
-        while (squeezedIdx > 0 && historyIdx > 0) {
-            squeezedIdx--;
-            historyIdx--;
-            DripsHistory memory drips = dripsHistory[historyIdx];
+        for (uint256 i = 1; i <= dripsHistory.length && i <= currCycleConfigs; i++) {
+            DripsHistory memory drips = dripsHistory[dripsHistory.length - i];
             if (drips.receivers.length != 0) {
-                uint32 squeezeStart = nextSqueezed[squeezedIdx];
+                uint32 squeezeStart = nextSqueezed[currCycleConfigs - i];
```
Comment (xmxanuel) [inline src/Drips.sol:422]: The name always confuses me a bit. Because the `squeezeStart` is only the squeeze start for the first squeeze in a new cycle in the other cases it is the `updateTime`.   It is kind of a `potentialsqueezeStart`

File: src/Drips.sol [hunk 3]
```diff
@@ -359,34 +374,55 @@ abstract contract Drips {
...
                 if (squeezeStart < _currCycleStart()) squeezeStart = _currCycleStart();
                 if (squeezeStart < squeezeEnd) {
-                    squeezedIdxs[squeezedNum++] = squeezedIdx;
+                    squeezedRevIdxs[squeezedNum++] = i;
```
Comment (xmxanuel) [inline src/Drips.sol:425]: It is a pity that the actual size of `squeezedRevIdxs` is unkown and we first need to iterate backwards at once to figure it out.

Outer-wise, we could have constructed it in the correct way.

Comment (xmxanuel) [reply src/Drips.sol:425]: Maybe it make sense to to define the following variable

```solidity
historyHashIdx = currCycleConfigs - i;
```

```suggestion
                    squeezedRevIdxs[squeezedNum++] = historyHashIdx;
```

Comment (xmxanuel) [reply src/Drips.sol:425]: Because the same subtraction happens in the higher function again

Comment (xmxanuel) [reply src/Drips.sol:425]: Naming could be better because it is also the position in the `nextSqueezed[currCycleConfigs - i];`

**Review by xmxanuel — COMMENTED** [review]
Overall, looks good. 

I think my main question is:
How important is the order of the emitted event hashes?
It would simplify the code if we can live with a reverse order.

File: src/Drips.sol (hunk 1 above)
Comment (CodeSandwich) [reply src/Drips.sol:351]: Yes, and that's terrible. It took me way too long trying to tie all these things together and weird `squeezedRevIdxs` was the best solution I've found :shrug:

**Comment by CodeSandwich** [comment]
Emitting events in the reverse order would only simplify a single line:
```solidity
uint256 revIdx = squeezedRevIdxs[squeezedNum - i - 1];
```
would become:
```solidity
uint256 revIdx = squeezedRevIdxs[i];
```
I think that it's worth keeping, because it makes the API more elegant, more obvious and generally easier to handle.

File: src/Drips.sol (hunk 1 above)
Comment (xmxanuel) [reply src/Drips.sol:351]: Yes, it looks good. Just more an overall comment.

File: src/Drips.sol (hunk 3 above)
Comment (xmxanuel) [reply src/Drips.sol:425]: Ah true. You are right :+1:

**Review by xmxanuel — APPROVED** [review]
